File: js/commands/virt.js
/* ROOT OF CRIME – virsh / mount / lsblk / exit */

Object.assign(Terminal, {
  _cmdVirsh(args) {
    const sub = (args[0] || '').toLowerCase();
    const guests = (this.ctx && this.ctx.guests) || {};
    const names = Object.keys(guests);
    if (!sub || sub === 'help') {
      return this._ok('usage: virsh list | virsh console <guest> | virsh attach-disk <guest> <size> | virsh reboot <guest>');
    }
    if (sub === 'list') {
      const lines = [' Id    Name                           State', '----------------------------------------------------'];
      names.forEach((n, i) => {
        const g = guests[n];
        lines.push(' ' + String(i + 1).padEnd(5) + n.padEnd(31) + (g.state || 'running'));
      });
      if (!names.length) lines.push(' (no guests)');
      return this._ok(lines.join('\n'));
    }
    if (sub === 'console') {
      return this.attach(args[1] || Virt.GUEST);
    }
    if (sub === 'attach-disk' || sub === 'attach_disk') {
      const rest = args.slice(1);
      let name = Virt.GUEST;
      let sizeTok = '';
      rest.forEach((a) => {
        if (a === '--size' || a === '-s') return;
        if (guests[a]) name = a;
        else sizeTok = a.replace(/^--size=/, '');
      });
      const guest = Virt.get(this.ctx, name);
      if (!guest) return this._err('virsh: failed to get domain');
      const parsed = Virt.parseSize(sizeTok);
      if (!parsed.ok) return this._err('virsh: ' + parsed.error + ' (example: virsh attach-disk booking-vm 200M)');
      const r = Virt.attachVolume(guest, parsed.bytes);
      if (!r.ok) return this._err('virsh: ' + r.error);
      Virt.paint(guest);
      return this._ok("Disk attached to '" + guest.hostname + "' as /dev/sdb (" + sizeTok + ")");
    }
    if (sub === 'reboot') {
      return this._rebootGuest(args[1] || Virt.GUEST);
    }
    return this._err('virsh: command \'' + sub + '\' not found');
  },

  _rebootGuest(name) {
    const guest = Virt.get(this.ctx, name);
    if (!guest) return this._err('virsh: failed to get domain');
    if (this._attached === guest.id) this.detach();
    const r = Virt.reboot(guest);
    if (!r.ok) return this._err('virsh: ' + r.error);
    Virt.paint(guest);
    return this._ok("Domain '" + guest.hostname + "' is being rebooted");
  },

  _cmdReboot() {
    if (this._attached) {
      const guest = this.attachedGuest();
      Virt.reboot(guest);
      this.detach();
      Virt.paint(guest);
      return this._ok('The system is going down for reboot NOW!\nConnection closed.');
    }
    return this._err(
      'reboot: refusing to reboot the host. The last IT left a note about that.\n' +
      'Try:  virsh reboot booking-vm'
    );
  },

  _cmdLsblk() {
    const guest = this.attachedGuest();
    const lines = ['NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT'];
    if (!guest) {
      lines.push('sda      8:0    0  500M  0 disk');
      lines.push('└─sda1   8:1    0  500M  0 part /');
      return this._ok(lines.join('\n'));
    }
    guest.disks.forEach((d) => {
      const kb = Math.max(1, Math.round((d.total || 0) / 1024));
      lines.push(d.id.padEnd(6) + '  8:0    0  ' + String(kb + 'K').padStart(4) + '  0 disk');
      lines.push('└─' + (d.id + '1').padEnd(4) + ' 8:1    0  ' + String(kb + 'K').padStart(4) + '  0 part ' + (d.mount || ''));
    });
    return this._ok(lines.join('\n'));
  },

  _cmdMount(args) {
    const guest = this.attachedGuest();
    if (!args.length) {
      if (!guest) return this._ok('/dev/sda1 on / type ext4 (rw,relatime)');
      return this._ok(guest.disks.filter((d) => d.mount).map((d) => (
        d.device + ' on ' + d.mount + ' type ext4 (rw,relatime)'
      )).join('\n'));
    }
    if (!guest) return this._err('mount: only the guest has extra disks. virsh console first.');
    if (args.length < 2) return this._err('mount: usage: mount DEVICE DIR');
    const r = Virt.mount(guest, args[0], args[1]);
    if (!r.ok) return this._err(r.error);
    Virt.paint(guest);
    return this._ok('');
  },

  _cmdUmount(args) {
    const guest = this.attachedGuest();
    if (!guest) return this._err('umount: nothing to unmount on the host');
    if (!args[0]) return this._err('umount: usage: umount DIR|DEVICE');
    const r = Virt.umount(guest, args[0]);
    if (!r.ok) return this._err(r.error);
    Virt.paint(guest);
    return this._ok('');
  },

  _cmdSsh(args) {
    if (typeof Infra !== 'undefined') Infra.boot();
    const raw = (args[0] || '').trim();
    if (!raw || raw.startsWith('-')) {
      const listed = (typeof Infra !== 'undefined' ? Infra.known() : [])
        .filter((h) => h.id !== 'closet')
        .map((h) => '  ssh ' + h.id.padEnd(16) + h.role)
        .join('\n');
      return this._err('usage: ssh [user@]HOST\n' + (listed || '  ssh precinct-13\n  ssh booking-vm'));
    }
    const id = typeof Infra !== 'undefined' ? Infra.resolve(raw) : raw.replace(/^.*@/, '').toLowerCase();
    if (id === 'closet' || id === 'localhost' || id === '127.0.0.1') {
      if (!this._stack || !this._stack.length) return this._ok('already on closet');
      if (this.dropToDesk) this.dropToDesk();
      else while (this._stack && this._stack.length) this.popSession();
      return this._ok('Connection closed.\n' + (typeof VFS !== 'undefined' && VFS.closetMotd ? VFS.closetMotd() : ''));
    }
    const sess = typeof Infra !== 'undefined' ? Infra.sessionFor(id) : null;
    if (!sess) {
      return this._err('ssh: Could not resolve hostname ' + raw + '\nTry:  ssh precinct-13   or   ssh booking-vm');
    }
    if (sess.kind === 'guest') {
      return this.attach(sess.id || sess.host, { ctx: sess.ctx });
    }
    if (this._attached || this.host === 'booking-vm') this.detach();
    if (this.host === sess.host && this._remote === sess.host) {
      this.vfs = sess.vfs;
      this.ctx = sess.ctx;
      this.cwd = sess.cwd || sess.home || '/home/itguy';
      this.home = sess.home || this.home;
      this.user = sess.user || this.user;
      this._updatePrompt();
      return this._ok('already on ' + sess.host);
    }
    const ticket = typeof Game !== 'undefined' ? Game.ticketSession : null;
    this.pushSession({
      vfs: sess.vfs,
      cwd: sess.cwd || sess.home || '/home/itguy',
      host: sess.host,
      home: sess.home || '/home/itguy',
      user: sess.user || 'root',
      ctx: sess.ctx,
      attached: null,
      remote: sess.host,
      missionId: (ticket && ticket.id) || this.missionId
    });
    if (ticket && ticket.id) this.missionId = ticket.id;
    return this._ok(typeof Infra !== 'undefined' ? Infra.loginBanner(sess) : 'Last login: from closet');
  },

  _cmdExit() {
    if (this._stack && this._stack.length) {
      const r = this.detach();
      if (this.host === 'closet' && !this._remote && typeof VFS !== 'undefined' && VFS.closetMotd) {
        const closed = (r.stdout || '').replace(/\n$/, '');
        return this._ok(closed + '\n' + VFS.closetMotd());
      }
      return r;
    }
    return this._ok('logout');
  }
});

Terminal.define('ssh', function (args) { return this._cmdSsh(args); });
Terminal.define('virsh', function (args) { return this._cmdVirsh(args); });
Terminal.define('lsblk', function () { return this._cmdLsblk(); });
Terminal.define('mount', function (args) { return this._cmdMount(args); });
Terminal.define(['umount', 'unmount'], function (args) { return this._cmdUmount(args); });
Terminal.define('reboot', function () { return this._cmdReboot(); });
Terminal.define('exit', function () { return this._cmdExit(); });

File: crates/roc-kernel/src/shell/commands/virt.rs
//! virsh / lsblk / mount / reboot. Matches `js/commands/virt.js` minus ssh/exit.

use crate::guest::Guest;
use crate::virt as vengine;

use super::super::exec::Shell;
use super::super::output::Output;

impl Shell {
    pub(crate) fn cmd_virsh(&mut self, args: &[String]) -> Output {
        let sub = args.first().map(|s| s.to_ascii_lowercase()).unwrap_or_default();
        let names: Vec<String> = self.ctx.guests.keys().cloned().collect();
        if sub.is_empty() || sub == "help" {
            return Output::ok(
                "usage: virsh list | virsh console <guest> | virsh attach-disk <guest> <size> | virsh reboot <guest>",
            );
        }
        if sub == "list" {
            let mut lines = vec![
                " Id    Name                           State".to_string(),
                "----------------------------------------------------".to_string(),
            ];
            if names.is_empty() {
                lines.push(" (no guests)".into());
            } else {
                for (i, n) in names.iter().enumerate() {
                    let state = self
                        .ctx
                        .guests
                        .get(n)
                        .map(|g| g.state.as_str())
                        .unwrap_or("running");
                    lines.push(format!(" {:<5}{:<31}{state}", i + 1, n));
                }
            }
            return Output::ok(lines.join("\n"));
        }
        if sub == "console" {
            let name = args
                .get(1)
                .cloned()
                .unwrap_or_else(|| Guest::BOOKING.to_string());
            let Some(sess) = self
                .session_from_guest(&name)
                .or_else(|| self.infra.session_for(&name))
            else {
                return Output::err("ssh: Could not resolve hostname ".to_string() + &name);
            };
            if sess.kind != crate::infra::SessionKind::Guest {
                return Output::err("virsh: failed to get domain");
            }
            return self.attach_guest(sess);
        }
        if sub == "attach-disk" || sub == "attach_disk" {
            let rest = &args[1..];
            let mut name = Guest::BOOKING.to_string();
            let mut size_tok = String::new();
            for a in rest {
                if a == "--size" || a == "-s" {
                    continue;
                }
                if self.ctx.guests.contains_key(a) {
                    name = a.clone();
                } else {
                    size_tok = a.trim_start_matches("--size=").to_string();
                }
            }
            let parsed = match vengine::parse_size(&size_tok, None) {
                Ok(b) => b,
                Err(e) => {
                    return Output::err(format!(
                        "virsh: {e} (example: virsh attach-disk booking-vm 200M)"
                    ));
                }
            };
            let r = self.with_guest_mut(&name, |g| vengine::attach_volume(g, parsed));
            match r {
                None => Output::err("virsh: failed to get domain"),
                Some(Err(e)) => Output::err(format!("virsh: {e}")),
                Some(Ok(())) => Output::ok(format!(
                    "Disk attached to '{name}' as /dev/sdb ({size_tok})"
                )),
            }
        } else if sub == "reboot" {
            let name = args
                .get(1)
                .cloned()
                .unwrap_or_else(|| Guest::BOOKING.to_string());
            self.reboot_guest(&name)
        } else {
            Output::err(format!("virsh: command '{sub}' not found"))
        }
    }

    fn reboot_guest(&mut self, name: &str) -> Output {
        if self.attached.as_deref() == Some(name) {
            self.detach();
        }
        match self.with_guest_mut(name, |g| vengine::reboot(g)) {
            None => Output::err("virsh: failed to get domain"),
            Some(Err(e)) => Output::err(format!("virsh: {e}")),
            Some(Ok(())) => Output::ok(format!("Domain '{name}' is being rebooted")),
        }
    }

    pub(crate) fn cmd_reboot(&mut self) -> Output {
        if let Some(id) = self.attached.clone() {
            let _ = self.with_guest_mut(&id, |g| vengine::reboot(g));
            self.detach();
            return Output::ok("The system is going down for reboot NOW!\nConnection closed.");
        }
        Output::err(
            "reboot: refusing to reboot the host. The last IT left a note about that.\n\
             Try:  virsh reboot booking-vm",
        )
    }

    pub(crate) fn cmd_lsblk(&mut self) -> Output {
        let mut lines = vec!["NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT".to_string()];
        if let Some(id) = self.attached.clone() {
            if let Some(rows) = self.with_guest_mut(&id, |g| {
                g.disks
                    .iter()
                    .map(|d| {
                        let kb = ((d.total as f64) / 1024.0).round().max(1.0) as u64;
                        let size = format!("{kb}K");
                        format!(
                            "{:<6}  8:0    0  {:>4}  0 disk\n└─{:<4} 8:1    0  {:>4}  0 part {}",
                            d.id,
                            size,
                            format!("{}1", d.id),
                            size,
                            d.mount.clone().unwrap_or_default()
                        )
                    })
                    .collect::<Vec<_>>()
                    .join("\n")
            }) {
                lines.push(rows);
                return Output::ok(lines.join("\n"));
            }
        }
        lines.push("sda      8:0    0  500M  0 disk".into());
        lines.push("└─sda1   8:1    0  500M  0 part /".into());
        Output::ok(lines.join("\n"))
    }

    pub(crate) fn cmd_mount(&mut self, args: &[String]) -> Output {
        if args.is_empty() {
            if let Some(id) = self.attached.clone() {
                if let Some(text) = self.with_guest_mut(&id, |g| {
                    g.disks
                        .iter()
                        .filter(|d| d.mount.is_some())
                        .map(|d| {
                            format!(
                                "{} on {} type ext4 (rw,relatime)",
                                d.device,
                                d.mount.as_deref().unwrap_or("")
                            )
                        })
                        .collect::<Vec<_>>()
                        .join("\n")
                }) {
                    return Output::ok(text);
                }
            }
            return Output::ok("/dev/sda1 on / type ext4 (rw,relatime)");
        }
        let Some(id) = self.attached.clone() else {
            return Output::err("mount: only the guest has extra disks. virsh console first.");
        };
        if args.len() < 2 {
            return Output::err("mount: usage: mount DEVICE DIR");
        }
        match self.with_guest_mut(&id, |g| vengine::mount(g, &args[0], &args[1])) {
            None => Output::err("mount: only the guest has extra disks. virsh console first."),
            Some(Err(e)) => Output::err(e),
            Some(Ok(())) => Output::ok(""),
        }
    }

    pub(crate) fn cmd_umount(&mut self, args: &[String]) -> Output {
        let Some(id) = self.attached.clone() else {
            return Output::err("umount: nothing to unmount on the host");
        };
        let Some(target) = args.first() else {
            return Output::err("umount: usage: umount DIR|DEVICE");
        };
        match self.with_guest_mut(&id, |g| vengine::umount(g, target)) {
            None => Output::err("umount: nothing to unmount on the host"),
            Some(Err(e)) => Output::err(e),
            Some(Ok(())) => Output::ok(""),
        }
    }
}
